packages/server/src/koaJsonRpc/index.ts

```diff
@@ -122,10 +122,17 @@ export default class KoaJsonRpc {
         !hasOwnProperty(body, 'id') ||
         ctx.request.method !== 'POST'
       ) {
-        ctx.body = jsonResp(body.id || null, new InvalidRequest(), undefined);
-        ctx.status = 400;
-        ctx.state.status = `${ctx.status} (${INVALID_REQUEST})`;
-        return;
+        this.logger.warn(`Invalid request, requestIdIsOptional: ${this.requestIdIsOptional}, process.env.REQUEST_ID_IS_OPTIONAL: ${process.env.REQUEST_ID_IS_OPTIONAL}, body.jsonrpc: ${body.jsonrpc}, body[method]: ${body.method}, body[id]: ${body.id}, ctx.request.method: ${ctx.request.method}`);
+        if (this.requestIdIsOptional && !hasOwnProperty(body, 'id'))
+        {          
+          // If the request is invalid, we still want to return a valid JSON-RPC response, default id to 0
+          body.id = '0';
+        } else {
+          ctx.body = jsonResp(body.id || null, new InvalidRequest(), undefined);
+          ctx.status = 400;
+          ctx.state.status = `${ctx.status} (${INVALID_REQUEST})`;
+          return;
+        }
       }
```

Comment:
I would recommend extracting validation for `id` into another method for simplicity, and separate it from `POST` and other fields validation for clarity, just cause otherwise the lo added is out of context s and double evaluation of !hasOwnProperty(body, 'id') might not be needed.

Reply:
Agreed. Pulled it out into `hasInvalidReqestId()`